docs/RFCS/20200811_non_blocking_txns.md

```diff
@@ -0,0 +1,1027 @@
+* Route present-time reads to followers in consistent read Ranges
+  * Similar to existing follower read logic, but Range specific
+    * Add bit on RangeDescriptor? Or to Lease? Both are cached in client
+* Introduce [long-lived learner replicas](https://github.com/cockroachdb/cockroach/issues/51943)
+  * Biggest work item, but...
+  * Already being worked on and generally useful outside of this proposal
+  * Also, not a hard requirement for the rest of this proposal
+
+#### Tuning non_blocking_duration
+
+The `non_blocking_duration` is the distance in the future at which non-blocking
+transactions acquire locks and write provisional values. The non-blocking
+property of these transactions is provided when this is far enough in the future
+such that a non-blocking transaction is able to navigate its entire operation
+set, commit, and resolve intents before its commit timestamp its passed by
+"present time + max clock offset". Further, the non-blocking property of these
+transaction is provided on followers when all of these effects are replicated
+and applied on followers by the time that the non-blocking transaction's commit
+timestamp is passed by "present time + max clock offset".
+
+As such, some tuning of this value will need to be performed, likely taking into
+account transaction latency and replication latency.
+
+#### Non-Blocking Transaction Pushing
+
+A difficulty we face with tuning the `non_blocking_duration` is that we don't
+know how long a non-blocking transaction is going to take between the time that
+it writes its first intent until the time that it commits and resolves all of
+its intents. This means that if we tune too low, the transaction's intents may
+not be resolved by the time they drop below "present time" + uncertainty (i.e.
+become visible to readers) and this could cause unintentional blocking of
+readers. A potential mitigation to this is to actively push these intents
+forward using a mechanism similar to the `rangefeedTxnPusher`. This mechanism
+would monitor active intents on consistent read Ranges and push them and their
+transaction forward if they ever got too close to present time without being
+resolved.
+
+Such a mechanism would also help ensure that even if the coordinator for a
+non-blocking transaction died, it would be cleaned up before a read on one of
+the followers conflicted with it and was forced to perform a WAN RTT.
+
+An alternative approach that would have a similar effect is to buffer a
+non-blocking transactions writes in its coordinator until it is ready to commit.
+This would mean that we would only need to tune `non_blocking_duration` to the
+expected commit latency of a transaction, instead of its expected evaluation and
+commit latency combined.
+
+We'll probably want to do one or both of these approaches.
+
+## Drawbacks
+
+Complexity. But not overly so, especially compared to Consistent Read Replicas.
+
+Only effective if we can reduce the clock uncertainty interval dramatically.
```

Comment:
This is the biggest question in my mind. Is this achievable? If we restrict to CC then I suspect it is, but is that an acceptable restriction given that many of our current customers who want this are on-prem?

Reply:
We know that <1ms latencies are achievable, according to the experts at TickTock Networks. But even times <30ms would make this proposal compelling. And even <100ms is useable, IMO, since that's down to O(RTT). I think if we built clock sync'ing into the KV nodes themselves, we could easily achieve these timings even for on-prem. All that said, I don't think we should be making on-prem support a prerequisite for accepting/rejecting this proposal. We've said we're "cloud-first", and this is a perfect example of what that means in practice. It means we'll pick the very best technical proposal for cloud, even if it's harder (or even impossible) to implement for on-prem. Of course, many of our current customers are on-prem today. But we're building for orders of magnitude more customers we expect to be on cloud in the future.

Reply:
I didn't mean to imply that on-prem support should be a prerequisite. I'd just like to have the solution here fleshed out a bit more (perhaps in a companion doc), both for CC and for on-prem. The on-prem fleshing out could be: "we don't support this".

Reply:
I feel similarly to Andy about both the feasibility of improving clock synchronization and the applicability of this performance optimization even if it was to be restricted to CC.

But I'll also add that we are currently very conservative with our clock skew limits, which I attribute to our full deferral of clock synchronization to the black box of NTP. For a database that relies on synchronized clocks, we really haven't pushed very hard on time synchronization. We don't measure clock drift (rate), only clock skew (absolute). We don't measure time between synchronization and extrapolate a maximum possible skew given a max clock drift. And we certainly don't do anything more sophisticated like TickTock was doing. Even without all that, I still get the sense that we could do better here. A similarly architected system that does nothing more than CockroachDB does to maintain clock sync sets their default clock skew bounds a [full order of magnitude lower than we do](https://github.com/yugabyte/yugabyte-db/blob/3b90e8560b8d8bc81fba6ba9b9f2833e83e2244e/src/yb/util/physical_time.cc#L36-L38).